persistent/Database/Persist/Sql.hs

```diff
@@ -74,3 +95,17 @@ transactionUndoWithIsolation isolation = do
     conn <- ask
     let getter = getStmtConn conn
     liftIO $ connRollback conn getter >> connBegin conn getter (Just isolation)
+
+
+-- | Run a database action within a transaction
+--
+-- @since 2.10.2
+withTransaction :: MonadIO m => ReaderT SqlBackend m a -> ReaderT SqlBackend m a
+withTransaction action = do
+    conn <- ask
+    let getter = getStmtConn conn
+    liftIO $ connBegin conn getter Nothing
+    a <- action
+    liftIO $ connCommit conn getter
+    pure a
```

Comment:
So this isn't exception-safe. We'll get a dangling transaction if `action` throws. The function `runSqlConn` does this right:

```haskell
runSqlConn :: (MonadUnliftIO m, BackendCompatible SqlBackend backend) => ReaderT backend m a -> backend -> m a
runSqlConn r conn = withRunInIO $ \runInIO -> mask $ \restore -> do
    let conn' = projectBackend conn
        getter = getStmtConn conn'
    restore $ connBegin conn' getter Nothing
    x <- onException
            (restore $ runInIO $ runReaderT r conn)
            (restore $ connRollback conn' getter)
    restore $ connCommit conn' getter
    return x
```

But, if you're using this function (which you should be), then you're always going to be in a transaction. So doing `BEGIN` here would make a nested transaction, which is not really well supported in many databases, and likely not what is desired.